tests/benchdnn/graph/input_displacer.cpp

```diff
@@ -136,6 +136,120 @@ partition_data_displacer_t::partition_data_displacer_t(
                 }
             }
         }
+
+        // Alternatively, looking for Add->SoftMax chain, which represents
+        // explicit SDPA mask, and should be filled with upper-corner with -inf:
+        // 0 -inf -inf -inf
+        // 0    0 -inf -inf
+        // 0    0    0 -inf
+        // 0    0    0    0
+        // This is done to avoid taking future tokens into account by
+        // influencing SoftMax input values.
+        while (aop.kind_ == "Add" || aop.kind_ == "Select") {
+            auto *aop_out_lt = &aop.out_lts_[0];
+            auto *child_op = &dg_->get_op_by_in_lt(aop_out_lt->id_);
+            if (child_op->kind_ != "SoftMax") break;
+
+            // Softmax must be a part of same partition as the mask. This is to
+            // avoid cases, where mask is the last op in the partition, from
+            // being modified.
+            if (op_ids_set_.find(child_op->id_) == op_ids_set_.end()) break;
+
+            // Search for an input lt without a parent, this is the one to
+            // modify for both explicit and implicit masks.
+            const deserialized_lt *causal_mask_lt = nullptr;
+            size_t offset = SIZE_MAX;
+            size_t qk_data_offset = SIZE_MAX;
+            // Need to iterate over all inputs to handle padding mask expressed
+            // through Select op.
+            for (size_t i = 0; i < aop.in_lts_.size(); i++) {
+                auto *aop_in_lt = &aop.in_lts_[i];
+                auto *parent_op = &dg_->get_op_by_out_lt(aop_in_lt->id_);
+                if (!parent_op->empty()) {
+                    if (aop_in_lt->get_data_type()
+                            != logical_tensor::data_type::boolean) {
+                        // Filtering out condition for implicit causal mask.
+                        //
+                        // This is the qk_data, need to know its offset to
+                        // properly fill condition for padding mask.
+                        qk_data_offset = i;
```

Comment:
IIUC, The logic here wants to make sure that an implicit causal mask should not encounter this condition.
However, It seems it can still run into this condition:
1. The src_0 (qk_data) of the select operation in implicit causal mask has a parent_op (which passes the first condition of `!parent_op->empty()`).
2. The data type of src_0 is not boolean (which passes the second condition of `aop_in_lt->get_data_type() != logical_tensor::data_type::boolean`).

So it seems possible for the implicit causal mask to encounter `qk_data_offset = i`, which will make the data filling later incorrect.

Reply:
Agreed, thanks for testing it for me. I've introduced an additional variable to track whether the condition tensor had a parent or not, it should resolve it.